### PR TITLE
[inductor][cpp][gemm] optimize arbitrary N in packed gemm template

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130821
* #131024
* #129348
* __->__ #130690
* #130675

Currently we require `n % register_block_n == 0` which typically bring good perf when `n` is a multiply of 8, 16, 32 etc. while will fall back to the reference micro gemm otherwise (where `register_block_n == 1`). This PR optimizes this by padding `n` to the multiple of `register_block_n` which is 8, 16, 32 etc. for packed weight. Therefore, the micro-gemm can work as is on the padded `n`. When the weight is padded, we will use the local accumulation buffer to get the result from micro-gemm and then unpadded (sliced) before storing back to the output buffer.

Performance numbers measured on "Intel (R) Xeon (R) CPU Max 9480", single core, bf16.

Before
AUTOTUNE linear_unary(512x768, 3073x768, 3073)
  _linear_pointwise 2.3563 ms 100.0% 
  cpp_packed_gemm_0 710.5902 ms 0.3%

After
AUTOTUNE linear_unary(512x768, 3073x768, 3073)
  cpp_packed_gemm_0 1.8909 ms 100.0% 
  _linear_pointwise 2.1016 ms 90.0%

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 1]
```diff
@@ -121,9 +119,15 @@
                         {{ micro_gemm.codegen_call(kernel, tile_X, tile_W, acc, accum=True)|indent(24, false) }}
                     }
                 }
-                {%- set tile_Y = kernel.slice_nd(Y_2d, [("m_start", "m_end"), ("n_start", "n_start + N0")]) %}
+                {%- if N == PADDED_N %}
+                    {%- set tile_Y = kernel.slice_nd(Y_2d, [("m_start", "m_end"), ("n_start", "n_start + N0")]) %}
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp_gemm_template.py:123]: Can this line also be `{%- set tile_Y = kernel.slice_nd(Y_2d, [("m_start", "m_end"), ("n_start", "n_end")]) %}`, so we can share same line as implementation of `padded_n`?

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -121,9 +119,15 @@
...
+                    {%- set tile_acc = acc %}
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp_gemm_template.py:124]: Feel like this line can also be `{%- set tile_acc = kernel.slice_nd(acc, [(), ("0", "n_end - n_start")]) %}`? For no_paddeding case, will it be same as `acc`?

**Review by leslie-fang-intel — APPROVED** [review]
LGTM, feels we can save some lines in the template.

File: torch/_inductor/codegen/cpp_template_kernel.py [hunk 3]
```diff
@@ -281,7 +281,7 @@ def store_output(
                     scope.add_local_buffer(src)
                     return self.store_pointwise_nodes(dst, [copy])
             else:
-                assert dst.layout == src.layout
+                assert dst.layout == src.layout, f"dst: {dst}, src: {src}"
```
Comment (Skylion007) [inline torch/_inductor/codegen/cpp_template_kernel.py:284]:
```suggestion
                assert dst.layout == src.layout, f"{dst=}, {src=}"

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp_gemm_template.py:123]: For cases without padding, I don't want the expressions depend on `n_end` which is runtime computed but on the constant `N0`. That's why I used the conditionals...

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 2 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp_gemm_template.py:124]: ditto

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**